src/Fhir.Anonymizer.Core/AnonymizerEngine.cs

```diff
@@ -117,29 +126,52 @@ public ElementNode AnonymizeResourceNode(ElementNode root)
 
                 foreach (var node in matchedNodes)
                 {
-                    AnonymizeChildNode(node, rule, resourceContext.PathSet, resourceId);
+                    AnonymizeChildNode(node, rule, resourceContext.PathSet, resourceId, status);
                 }
             }
 
             return root;
         }
 
-        private void AnonymizeInternalResourceNodes(List<ElementNode> resourceNodes)
+        private void AnonymizeContainedResourceNodes(List<ElementNode> resourceNodes, AnonymizationStatus status)
+        {
+            // For contained, all contained resources share the same status with root resource
+            foreach (var resource in resourceNodes)
+            {
+                var newResource = AnonymizeResourceNode(GetResourceRoot(resource), status);
+                resource.Parent.Replace(_provider, resource, newResource);
+            }
+        }
+
+        private AnonymizationStatus AnonymizeBundleEntryResourceNodes(List<ElementNode> resourceNodes)
         {
-            foreach(var resource in resourceNodes)
+            var bundleStatus = new AnonymizationStatus();
+            // For Bundle, every entry has its own status
+            foreach (var resource in resourceNodes)
             {
-                var newResource = AnonymizeResourceNode(GetResourceRoot(resource));
+                var status = new AnonymizationStatus();
+                var anonymizedNode = AnonymizeResourceNode(GetResourceRoot(resource), status);
+                
+                anonymizedNode.RemoveNullChildren();
+                var anonymizedResource = anonymizedNode.ToPoco<Resource>();
+                
+                anonymizedResource.TryAddSecurityLabels(status);
```

Comment:
For a bundle, do we add security label to Bundle resource or each entry resource?

Reply:
For a Bundle, if entry resources are anonymized, security labels will both be added to Bundle resource and entry resources. We did not find any specific guidance for this scenario in both the specs of [security labels](https://www.hl7.org/fhir/security-labels.html) and [Bundle](http://hl7.org/fhir/bundle.html). But we find that it is OK to add security labels to Bundle resource (in the description of both spec) and entry resources (in an [example](https://www.hl7.org/fhir/security-labels.html#rsl) of the spec of security labels). Here are some [internal discussion](https://microsofthealth.visualstudio.com/Health/_workitems/edit/72972) around it in the discussion field of this working item.